Recursively Delete Unreferenced Index Directories

original-brownbear opened the pull request:
* Add ability to list child "folders" in the blob store
   * Use it to implement recursive delete on all stale index folders when cleaning up instead of using the diff between two `RepositoryData` instances to cover aborted deletes and blocked primaries writing things after a snapshot has already failed
   * The safety of this operation is guaranteed by the fact that we get the list of to-delete indices before we write the new index-N blob, hence we can't accidentally delete any indices that might be created after the index-N has been written and will only run the cleanup if writing the index-N succeeds.
* Clean up on delete operations to keep the number of blobs to clean up in continuous operation low. In case of a clean repository the overhead of running the cleanup logic is completely negligible (it's only a few list calls on the indices path that's it and those are cheap since we can and do filter for first level children only in all three implementations).
* Add new infrastructure for running `*IT` style third party tests against real Azure, GCS and S3 repositories.
   * For S3 we also run the tests against Minio
   * I manually verified the 3rd party tests to work with real repositories and they don't require new infrastructure in Jenkins since they are triggered via the same env variables used by the REST style 3rd party tests already. The performance on GCS and S3 seems to be a non-issue due to bulk deletes, on Azure the situation is a little more dicey and since we don't have bulk delete and I'll try to speed up deletes there in another PR I'm working on now. 

* closes #13159

original-brownbear committed:
Recursively Delete Unreferenced Index Directories

* Add ability to list child "folders" in the blob store
   * Use it to implement recursive delete on all stale index folders when cleaning up instead of using the diff between two `RepositoryData` instances to cover aborted deletes

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear committed:
more bulk

original-brownbear committed:
add docs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear commented on server/src/test/java/org/elasticsearch/snapshots/MetadataLoadingDuringSnapshotRestoreIT.java:
```diff
@@ -141,8 +141,8 @@ public void testWhenMetadataAreLoaded() throws Exception {
         assertGlobalMetadataLoads("snap", 1);
-        assertIndexMetadataLoads("snap", "docs", 5);
-        assertIndexMetadataLoads("snap", "others", 4);
+        assertIndexMetadataLoads("snap", "docs", 4);
```
Now we don't need to read meta blobs to get an index's shard count before deleting since we just get that from the child container listing.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -426,46 +425,75 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
-                repositoryData = getRepositoryData();
+                final RepositoryData repositoryData = getRepositoryData();
                 updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
+                // Cache the indices that were found before writing out the new index-N blob so that a stuck master will never
```
I think this approach of listing before writing the updated `index-N` makes this approach completely safe even on an eventually consistent blob store. If an index uuid (the snapshot uuid, not the CS one!) goes out of scope here for a single `index-N` it will never be reused in a future `index-M` (`M > N)` => we don't have to care about stuck master nodes coming back to haunt us.
What we could however do is optimize this a little and maybe write a `SUCCESS-N` blob after we finished all operations that went into a given `index-N` blob and only run the listing if we suspect dangling indices because an operation that wrote a given `index-N` didn't finish fully.
The downside of this is, that it gets a little complicated when you have to account for master failover scenarios and may not be worth it performance wise now that we have bulk deletes and could parallelize the recursive delete as well.

original-brownbear commented:
@ywelsch @andrershov I think this would be a very neat low-risk step we could take independently of other improvements to the blob store. This will just automatically clean up stale indices down the line and there is no need of any tomb-stone logic with this solution.
Going ahead with this would only leave cleaning up stale `snap-${uuid}.dat` files (and subsequently fixing the related `index-N` in the shard folders) and we'd be good on the stale blobs front I think (this is a little trickier but also less important in it's likely impact on repository size).

Just asking for some early feedback here on the WIP PR before I go through the pain of implementing actual integration tests that verify the list children implementations :)

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -426,46 +425,75 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+            final var survivingIndices = updatedRepositoryData.getIndices();
             deleteIndices(
                 Optional.ofNullable(finalSnapshotInfo)
-                    .map(info -> info.indices().stream().map(repositoryData::resolveIndexId).collect(Collectors.toList()))
```
No need to run partial deletes on indices that went fully out of scope now. This would be great to have for #41581 because if we were to delay writing the metadata in the current implementation, we run into the messy spot of not knowing about the shards if we fail to write the index metadata -> current implementation couldn't delete the stale blobs from a partial upload of a new index.

original-brownbear commented:
@ywelsch @andrershov I take todays discussion as meaning that the general idea here is good enough so I'll go ahead and implement strong tests here + clean up the code in the repository implementations => no need to review this yet, I'll re-request review once ready (should be shortlyish).

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
cleanup after snapshotting as well

original-brownbear committed:
fix checkstyle

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
nicer

original-brownbear committed:
start marker logic

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
clean

original-brownbear committed:
prevent index dir blobs from being listed

original-brownbear committed:
add assertion about proper S3 behavior

original-brownbear committed:
fix azure blob children listing

original-brownbear committed:
fix azure

original-brownbear committed:
nicer

original-brownbear committed:
catch all the exceptions

original-brownbear committed:
cleanup

original-brownbear committed:
smaller changeset

original-brownbear committed:
nicer

original-brownbear committed:
nicer

original-brownbear committed:
dry up test infra

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear committed:
test cleanup on hdfs repository

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear commented:
@ywelsch @andrershov I think the implementation here is getting to an ok point now, but I'm wondering about the right approach to testing this.

It seems to me the only way of getting some real coverage on GCS and Azure here would be third party tests. Unfortunately, aside from just having the real repositories available we also need the ability to do a few operations (create some stale data and verify it's gone later on) which the current rest 3rd party tests don't easily allow for.

What do you think about doing the same thing I did for HDFS for tests for S3, Azure and GCs (see HDFS impl. https://github.com/elastic/elasticsearch/pull/42189/files#diff-c347d8684917892a49d79be15fca9678R165) and create 3rd party tests using the `*IT` infrastructure? (for HDFS this was zero effort since HDFS offered a proper mock impl.) This seems like the least code required option and easiest to maintain. If we implement these tests as REST tests then we'd need some secondary infrastructure for accessing the repositories which seems kind of horrible.

original-brownbear commented:
Jenkins run elasticsearch-ci/2

ywelsch commented:
> What do you think about... and create 3rd party tests using the *IT infrastructure? 

makes sense to me.

original-brownbear commented:
> >What do you think about... and create 3rd party tests using the *IT infrastructure?

> makes sense to me.

Alright will implement tomorrow :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
azure 3rd party cleanup tests

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
gcs third party tests

original-brownbear committed:
s3 third party tests

original-brownbear committed:
better comment

original-brownbear committed:
add todo about eventual consistency on S3

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
much drier

original-brownbear committed:
much drier consistency checks

original-brownbear committed:
Fix consistency tests to account for S3 eventual consistency

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
S3 resiliency tests work with minio

original-brownbear committed:
add comment on Minio hack

original-brownbear committed:
shorter diff

original-brownbear commented:
@ywelsch @andrershov I think this one is good to go. Added (third party) tests for the cleanup for all repository implementations now and cleaned things up reasonably.

Once Cloud gets around to providing some test data it would be good to verify the performance on Azure though (I did some tests on S3 and GCS and the time a cleanup takes is very reasonable here since we can delete 1k or 100 blobs per delete call respectively ... also, worst case someone aborting a cleanup operation forcefully isn't catastrophic and the next run of a snapshot or delete operation will simply pick up where the aborted one left off effectively).
I'll look into Azure performance now in parallel, it should be possible to speed things up there as well. The idea of potentially having to clean up tens of thousands of blobs sequentially seems a little problematic otherwise.

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear commented:
Jenkins run elasticsearch-ci/1 (slow Jenkins ...)

ywelsch commented:
Can you split out the infrastructure for the third-party tests here into a separate PR (and add a simple  test to show that it works)? It will make this easier to review as a follow-up then.

original-brownbear commented:
@ywelsch sure on it :)

original-brownbear commented:
@ywelsch done extracting the test infra in https://github.com/elastic/elasticsearch/pull/42586

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear commented:
@ywelsch this should be good for a review now, the extraction of the test infrastructure is merged into `master` and hence this PR now :)

ywelsch commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -231,6 +233,50 @@ public void deleteBlobIgnoringIfNotExists(String blobName) throws IOException {
+            }
+            return Maps.ofEntries(entries);
+        } catch (final AmazonClientException e) {
+            throw new IOException("Exception when listing children", e);
```
also mention current path here

ywelsch commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -231,6 +233,50 @@ public void deleteBlobIgnoringIfNotExists(String blobName) throws IOException {
+                } else {
+                    final ListObjectsRequest listObjectsRequest = new ListObjectsRequest();
+                    listObjectsRequest.setBucketName(blobStore.bucket());
+                    listObjectsRequest.setPrefix(keyPath);
```
In the example that the SDK provides for `getCommonPrefixes`, it adds `/` to the prefix that is queried. Should we do the same?

```
     * If calling <code>listObjects</code> with the prefix="foo/" and the delimiter="/" on 
     * this bucket, the returned <code>S3ObjectListing</code> 
     * will contain one entry in the common
     * prefixes list ("foo/bar/") and none of the keys beginning with that
     * common prefix will be included in the object summaries list.
```

original-brownbear replied:
As a matter of fact I think that Javadoc is wrong. If I add the `/` listing child directories fails in my testing. I can look into that after my vacation and maybe open a PR to the SDK.

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -158,6 +158,16 @@ default void deleteBlobIgnoringIfNotExists(String blobName) throws IOException {
 
+    /**
+     * Lists all child containers under this container. A child container is defined as a container whose {@link #path()} method returns
+     * a path that has the return  of this container's {@link #path()} method as a prefix with a single path component added to it.
```
this reads a bit funnily, maybe rephrase

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -73,6 +74,22 @@ public FsBlobContainer(FsBlobStore blobStore, BlobPath blobPath, Path path) {
+        try (DirectoryStream<Path> stream = Files.newDirectoryStream(path)) {
+            for (Path file : stream) {
+                final BasicFileAttributes attrs = Files.readAttributes(file, BasicFileAttributes.class);
+                if (attrs.isDirectory()) {
```
why not just use `Files.isDirectory(file)`?

original-brownbear replied:
jup ... this was just a left over from copying the logic of list by prefix, adjusting :)

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -73,6 +74,22 @@ public FsBlobContainer(FsBlobStore blobStore, BlobPath blobPath, Path path) {
 
+    @Override
+    public Map<String, BlobContainer> children() throws IOException {
+        // If we get duplicate files we should just take the last entry
```
why would we expect duplicates?

original-brownbear replied:
Good question, for some reason I just mindlessly copied that comment from the list by prefix method. It makes 0 sense imo, I'll just remove it.

ywelsch reviewed (changes requested):
I've only reviewed the `BlobContainer.children()` method. Can we have a separate PR for this and more extensive testing for that as well in the form of third-party tests? I'm thinking tests along the lines of `ESBlobStoreContainerTestCase`, but against the real service. We can use `AbstractThirdPartyRepositoryTestCase` for that, and get the `BlobContainer` directly from the `BlobStoreRepository` of the node.

original-brownbear commented:
@ywelsch sure on it, will fix the points you raised in that then :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -142,6 +142,22 @@ private boolean doesBucketExist(String bucketName) {
+        final String pathStr = path.buildAsString();
+        final MapBuilder<String, BlobContainer> mapBuilder = MapBuilder.newMapBuilder();
+        SocketAccess.doPrivilegedVoidIOException
+            (() -> client().get(bucketName).list(BlobListOption.currentDirectory()).iterateAll().forEach(blob -> {
```
🤦‍♂ yea, obviously when I fixed the listing style to use `BlobListOption.currentDirectory()` I just replaced the prefix condition with it ..., will fix.

original-brownbear commented:
@ywelsch extracted logic for listing children in https://github.com/elastic/elasticsearch/pull/42653 and either adjusted code or answered your points here I hope.

andrershov commented:
@original-brownbear could you please integrate changes made by the spin-off PR (seems that there are a lot of merge conflicts) and re-request review from me afterwards?

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
shorter diff

original-brownbear commented:
@andrershov @ywelsch this one is good to review again :) thanks!

ywelsch commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsTests.java:
```diff
@@ -145,6 +156,83 @@ public void testSimpleWorkflow() {
+        BlobStoreTestUtil.assertConsistency(repo, repo.threadPool().executor(ThreadPool.Names.GENERIC));
+    }
+
+    public void testCleanup() {
```
why have this test here? Isn't this covered by `HdfsRepositoryTests`?

original-brownbear replied:
Wops, leftover from extracting things to other PRs -> removing :)

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java:
```diff
@@ -418,9 +419,9 @@ public void testSnapshotDuringNodeShutdown() throws Exception {
         SnapshotInfo snapshotInfo = waitForCompletion("test-repo", "test-snap", TimeValue.timeValueSeconds(60));
         logger.info("Number of failed shards [{}]", snapshotInfo.shardFailures().size());
         logger.info("--> done");
+        BlobStoreTestUtil.assertRepoConsistency(internalCluster(), "test-repo");
```
instead of adding this to the end of each test, can't we just do this with an `@After`?

original-brownbear replied:
Can't do that unfortunately since we have some tests that leave the repo in a different broken state (I will explain below in a sec, but same goes for the other tests I added this to). Also, some tests use multiple repositories => adding it manually seemed easier than adding some other logic to keep track of what repos were created etc.

ywelsch replied:
should we instead have those tests opt out of the check behavior perhaps? It creates less noise and is perhaps more specific in what remains to be addressed.

original-brownbear replied:
I guess we could do that => let me see what I can do there, on it :)

original-brownbear replied:
How about af47bb8 ? :) Seems my memory didn't serve me well here and we actually only have a single test that leave the repo in broken (and not yet cleaned up state). :)

ywelsch replied:
> How about af47bb8 ?

I like this better.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java:
```diff
@@ -181,6 +184,86 @@ protected void doRun() throws Exception {
         }
     }
 
+    public void testCleanup() throws Exception {
```
should we also add another test which does a full-cluster restart during a snapshot  or a snapshot delete? This should result in repo garbage as well, and will be more realistic in the garbage that's created.

original-brownbear replied:
We wouldn't fully pass that test yet :( As discussed, we are still leaking `snap-${uuid}.dat` blobs even with this fix. So if we force a full restart we will eventually get into the state of leaving those files. This PR only cleans up stale indices.

That said: I have another branch that also deals with this case but it has this one (and a few other cleanup PRs I raised) as a dependency and we can look into that once we're done with the stale indices I guess? Then I'm all for testing full cluster restarts :)

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -399,46 +398,88 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
         }
     }
 
+    private void cleanupStaleIndices(BlobContainer indicesBlobContainer, Map<String, BlobContainer> foundIndices,
```
why pass both `foundIndices` and `survivingIndices` to this method?  Perhaps compute `staleIndices`,  which is a subset of `Map<String, BlobContainer> foundIndices`, directly on  the caller side?

original-brownbear replied:
I basically just kept it like this to keep things a little drier. I figured since the ordering of RPC calls is really important in the two methods that call this so that things don't run in the wrong order I'd have as little logic in those as possible outside of the RPC calls they trigger.

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -399,46 +398,88 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+                }
+            }
+            try {
+                indicesBlobContainer.deleteBlobsIgnoringIfNotExists(toDelete);
```
this is still there to clean-up directories, right? We're somewhat breaking abstractions here.
I wonder if we should provide a `deleteDirectoryBlob` method on BlobContainer that can be more efficiently be implemented by `FSBlobContainer` and `HDFSBlobContainer`, instead of now iterating over all files for those as well. Alternatively, we could have a `deleteAll` method on BlobContainer that completely removes the BlobContainer, including the dir that it's part of.

ywelsch replied:
It would also make `deleteContents`, which is further below, an implementation detail.

original-brownbear replied:
As far as the abstractions go: I'm wondering if that is still true in the real world. I just tested this and if I create a few objects in a folder and delete them, I still get the common prefix listed until I delete it (even in the S3 UI I can create an empty folder that then comes up as a common prefix).
Also S3 nowadays treats any object ending in `/` as a folder and doesn't record its size or metadata (see https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html). The same goes for Azure and GCS afaik.

That said, I think it's technically fine to make this logic an implementation detail of `BlobContainer`, I mainly liked the current approach for its simplicity in back porting :)
Practically, I'm also not sure we get much out of it (yet):
* For FS and HDFS a recursive/folder delete will still have to go through all the directories and list all the files and the delete one by one. `FileTreeWalker` might be a little more efficient here algorithmically but I don't think it practically matters for real world NFS latencies?
* For Azure + S3 implementations I could see a stronger case because we could massively optimize here by moving to non-blocking IO SDK releases and run listing and deleting in parallel (technically we could do this already in Azure where we now have the internal thread pool ...). Since we need the listing functionality publicly exposed anyway for the cleanup algorithm, maybe we could do that refactoring just in `7.x` and `8` and look into possible optimizations there and keep changes for `6.x` as short as possible?

original-brownbear replied:
Sorry forgot to add: having the container delete itself is a little tricky in the current implementation(s) of the container and would require more changes than this approach because the container doesn't always have "all the information" available to delete itself (it's all trivial changes but not sure we want that noise for 6.x when it's just making the code look a little nicer there?).

ywelsch replied:
>  if I create a few objects in a folder and delete them, I still get the common prefix listed until I delete it

perhaps we should test for this in AbstractThirdPartyRepositoryTestCase to check that all implementations guarantee this behavior

ywelsch replied:
> For FS and HDFS a recursive/folder delete will still have to go through all the directories and list all the files and the delete one by one.

I'm not sure about file systems in general, but for HDFS at least, doing a recursive delete on the folder results in a single RPC call to the namenode, whereas the file-by-file delete does many many sequential RPC calls, each with significant latency.

I think we should provide folder deletion methods at the BlobContainer level so that each  type can have an efficient implementation.

Let's create a separate predecessor PR for that.

original-brownbear replied:
Right .. I underestimated HDFS there, my bad -> on it with the predecessor PR.

> perhaps we should test for this in AbstractThirdPartyRepositoryTestCase to check that all implementations guarantee this behavior

Sounds good, I'll add that test in the next PR :)

original-brownbear replied:
I opened https://github.com/elastic/elasticsearch/pull/43281 now that has custom (and optimized) delete calls for all the implementations individually :)

ywelsch reviewed (changes requested):
I've done a first pass, leaving some comments.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
CR: removed redundant test

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
CR: add repo consistency check globally

original-brownbear committed:
Checkstyle

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -399,46 +398,88 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+            } catch (IOException e) {
+                logger.warn(() ->
+                    new ParameterizedMessage(
+                        "[{}] failed to clean up unreferenced index folders.", metadata.name()), e);
```
add folder names here as well?

ywelsch replied:
also, no need to add a full stop at the end of the log message (same for other messages you've added as well).

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
CR: remove . in ex messages

original-brownbear committed:
remove now redundant directory deletes

original-brownbear commented:
@ywelsch I merged in `master` including the container `delete` method from #43281 now and it simplified away the discussion points around directory deletion.
Also, I removed all the full stops I addd to new log/ex. messages, this should be good for another round, thanks!

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear commented:
Sorry, seems the API change messes with some of the mock repository tests. On it fixing that.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Fix test issues with MockRepository

original-brownbear commented:
Fixed the mock repository based tests -> this should be good for another review :) thanks!

original-brownbear commented:
Jenkins run elasticsearch-ci/1

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -393,46 +392,68 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
             // Delete snapshot from the index file, since it is the maintainer of truth of active snapshots
-            final RepositoryData repositoryData;
             final RepositoryData updatedRepositoryData;
+            final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
```
move this to the try block where it's used

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -393,46 +392,68 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+                }
+            }
+        } catch (Exception e) {
+            // TODO: We shouldn't be blanket catching and suppressing all exceptions here and instead handle them safely upstream.
```
add an `assert false` here?

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -494,9 +515,13 @@ public SnapshotInfo finalizeSnapshot(final SnapshotId snapshotId,
-            final RepositoryData repositoryData = getRepositoryData();
-            writeIndexGen(repositoryData.addSnapshot(snapshotId, blobStoreSnapshot.state(), indices), repositoryStateId);
+            writeIndexGen(updatedRepositoryData, repositoryStateId);
+            cleanupStaleIndices(foundIndices, survivingIndices);
```
as discussed, let's not do the clean-up as part of the snapshotting, but only during deletion.
Instead let's think about how we can add a way for the user to force a clean even if there are no snapshots to be deleted.

original-brownbear replied:
Ok, will revert that. Unfortunately this will also force doing something about the new repo consistency tests (those were reliant on the fact that the last snapshot taken would clean up the repo). I'll think of something there :)

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/AbstractSnapshotIntegTestCase.java:
```diff
@@ -65,6 +67,23 @@ public void assertConsistentHistoryInLuceneIndex() throws Exception {
+                .stream()
+                .map(RepositoryMetaData::name)
+                .forEach(name -> BlobStoreTestUtil.assertRepoConsistency(internalCluster(), name));
+        }
```
perhaps log the reason in the else branch here

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java:
```diff
@@ -723,6 +722,7 @@ public boolean clearData(String nodeName) {
     }
 
     public void testRegistrationFailure() {
+        disableRepoConsistencyCheck("This test does not create any data in the repository.");
```
no dots at the end of logging messages (same for other occurrences in this PR)

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -182,7 +182,8 @@ public void testBasicWorkFlow() throws Exception {
 
         logger.info("-->  creating repository");
-        assertAcked(client.admin().cluster().preparePutRepository("test-repo").setType("fs").setSettings(randomRepoSettings()));
+        final String repoName = "test-repo";
```
revert this, not necessary anymore

ywelsch reviewed (changes requested):
I've left a few more comments

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
CR: comments

original-brownbear committed:
CR: move container creation

original-brownbear commented:
thanks @ywelsch I addressed all points. I tackled the consistency check issue by deleting a random snapshot before doing the consistency check in the `After` blocks now. For the repos where I couldn't run a delete because they were some messy "mock" repos that blocked on writes I just turned off the check as well now for the time being since I can't run a delete on those.

ywelsch commented:
There's a relevant test failure on the PR

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear committed:
Remute test

original-brownbear commented:
@ywelsch thanks for pinging, obviously had to re-mute #39852 now that we don't clean up on snapshot :( I'll find another solution for that test in a follow up, it's easy with the infrastructure we have now.

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -450,6 +450,7 @@ private void cleanupStaleIndices(Map<String, BlobContainer> foundIndices, Map<St
             // TODO: We shouldn't be blanket catching and suppressing all exceptions here and instead handle them safely upstream.
             //       Currently this catch exists as a stop gap solution to tackle unexpected runtime exceptions from implementations
             //       bubbling up and breaking the snapshot functionality.
+            assert false;
```
add exception to message that is shown when assertion breaks?

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/AbstractSnapshotIntegTestCase.java:
```diff
@@ -75,7 +75,16 @@ public void assertRepoConsistency() {
+                    BlobStoreTestUtil.assertRepoConsistency(internalCluster(), name);
+                });
+        } else {
+            logger.info("--> skipped repo consistency checks because [{}]" ,skipRepoConsistencyCheckReason);
```
space after comma

ywelsch reviewed (approved):
LGTM

original-brownbear committed:
CR: comments

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into allow-listing-directories

original-brownbear commented:
thanks @ywelsch !
